### ISSUE TITLE
[SDK52] [expo-camera] Unable to build project due to swift concurrency error

### ISSUE THREAD

**Issue by Eingin** [expo/expo#32378]
### Minimal reproducible example

https://github.com/Eingin/expo-camera-build-reproduction

### What platform(s) does this occur on?

iOS

### Where did you reproduce the issue?

in a development build

### Summary

While trying to upgrade an existing project to SDK52 i ran into this build issue on iOS inside `expo-camera`

```
❌  (node_modules/expo-camera/ios/Current/CameraView.swift:795:13)

  793 |   private func createBarcodeScanner() -> BarcodeScanner {
  794 |     let scanner = BarcodeScanner(session: session, sessionQueue: sessionQueue)
> 795 |     scanner.setPreviewLayer(layer: previewLayer)
      |             ^ call to actor-isolated instance method 'setPreviewLayer' in a synchronous main actor-isolated context
  796 |     scanner.onBarcodeScanned = { [weak self] body in
  797 |       guard let self else {
  798 |         return


❌  (node_modules/expo-camera/ios/Current/CameraView.swift:796:13)

  794 |     let scanner = BarcodeScanner(session: session, sessionQueue: sessionQueue)
  795 |     scanner.setPreviewLayer(layer: previewLayer)
> 796 |     scanner.onBarcodeScanned = { [weak self] body in
      |             ^ actor-isolated property 'onBarcodeScanned' can not be mutated from the main actor
  797 |       guard let self else {
  798 |         return
  799 |       }
```

This looks due to the recent change within the `expo-camera` module to use swift concurrency #31900

### Environment

```text
expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.6.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.12.2 - ~/.nvm/versions/node/v20.12.2/bin/node
      Yarn: 1.22.22 - /opt/homebrew/bin/yarn
      npm: 10.5.0 - ~/.nvm/versions/node/v20.12.2/bin/npm
      Watchman: 2024.09.30.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.4, iOS 17.4, macOS 14.4, tvOS 17.4, visionOS 1.1, watchOS 10.4
    IDEs:
      Android Studio: 2024.1 AI-241.18034.62.2411.12169540
      Xcode: 15.3/15E204a - /usr/bin/xcodebuild
    npmPackages:
      expo: ~52.0.0-preview.5 => 52.0.0-preview.5
      expo-router: ~4.0.0-preview.3 => 4.0.0-preview.3
      react: 18.3.1 => 18.3.1
      react-dom: 18.3.1 => 18.3.1
      react-native: 0.76.0 => 0.76.0
      react-native-web: ~0.19.13 => 0.19.13
    npmGlobalPackages:
      eas-cli: 12.4.1
    Expo Workflow: bare
```

### Expo Doctor Diagnostics

```text
✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check npm/ yarn versions
✔ Check for issues with Metro config
✔ Check if the project meets version requirements for submission to app stores
✔ Check for app config fields that may not be synced in a non-CNG project
✔ Validate packages against React Native Directory package metadata
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check native tooling versions
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check that native modules use compatible support package versions for installed Expo SDK
✖ Check that packages match versions required by installed Expo SDK

Detailed check results:

The following packages should be updated for best compatibility with the installed expo version:
  expo-dev-client@5.0.0-preview.2 - expected version: ~5.0.0-preview.2
Your project may not work correctly until you install the expected versions of the packages.
Found outdated dependencies
Advice: Use 'npx expo install --check' to review and upgrade your dependencies.
```

**Comment by benjaminkomen** [comment]
Just had the same issue with [this EAS build](https://expo.dev/accounts/generac-clean-energy/projects/installer-app/builds/a4161616-c941-4b58-8fcb-c844661f6815). Interestingly enough there was no issues with local dev client build, but a Preview build on EAS fails.

**Comment by alanjhughes** [comment]
Hi @Eingin - What versions of expo-camera and xcode are you using? Does it include this #32353. I don't believe it does. I just built on main and have no compiler issues

**Comment by Eingin** [comment]
@alanjhughes i was getting this with the latest version `16.0.2`
I was also able to replicate the build issue on the Expo test repo in here `bare-expo`

I've made a pr that fixes the issue: #32379

Are you able to clone my reproduction repo and see if it occurs for you?

**Comment by alanjhughes** [comment]
@Eingin the only way I could see this causing an error is if you have Swift 6 as the language version

**Comment by Eingin** [comment]
@alanjhughes I was able to reproduce the error on EAS using the `macos-sonoma-14.4-xcode-15.3` image

```json
{
  "build": {
    "production": {
      "ios": {
        "image": "macos-sonoma-14.4-xcode-15.3"
      }
    }
  }
}
```

**Comment by b0iq** [comment]
Can be fixed with **`expo-camera@16.0.2.patch`**

```patch
diff --git a/ios/Current/BarcodeScanner.swift b/ios/Current/BarcodeScanner.swift
index 94f2025d52354023be8a7a52e5de9e59b4aece81..bcc6cb88f98b1efeaeec35858df36fd621fb1070 100644
--- a/ios/Current/BarcodeScanner.swift
+++ b/ios/Current/BarcodeScanner.swift
@@ -4,7 +4,7 @@ import AVFoundation
 let BARCODE_TYPES_KEY = "barcodeTypes"
 
 actor BarcodeScanner: NSObject, BarcodeScanningResponseHandler {
-  var onBarcodeScanned: (([String: Any]?) -> Void)?
+  private var onBarcodeScanned: (([String: Any]?) -> Void)?
   var isScanningBarcodes = false
 
   // MARK: - Properties
@@ -76,6 +76,11 @@ actor BarcodeScanner: NSObject, BarcodeScanningResponseHandler {
     }
   }
 
+
+  func setOnBarcodeScanned(_ onBarcodeScanned: @escaping ([String: Any]?) -> Void) {
+    self.onBarcodeScanned = onBarcodeScanned
+  }
+
   func maybeStartBarcodeScanning() async {
     guard isScanningBarcodes else {
       return
diff --git a/ios/Current/CameraView.swift b/ios/Current/CameraView.swift
index 2694c66bddf9272a2dd235816a8d74434bc3f601..e8c630668ccb5f5a9dee7feee0af22e6b2780320 100644
--- a/ios/Current/CameraView.swift
+++ b/ios/Current/CameraView.swift
@@ -792,13 +792,15 @@ public class CameraView: ExpoView, EXAppLifecycleListener,
 
   private func createBarcodeScanner() -> BarcodeScanner {
     let scanner = BarcodeScanner(session: session, sessionQueue: sessionQueue)
-    scanner.setPreviewLayer(layer: previewLayer)
-    scanner.onBarcodeScanned = { [weak self] body in
-      guard let self else {
-        return
-      }
-      if let body {
-        self.onBarcodeScanned(body)
+    Task {
+      await scanner.setPreviewLayer(layer: previewLayer)
+      await scanner.setOnBarcodeScanned { [weak self] body in
+        guard let self else {
+          return
+        }
+        if let body {
+          self.onBarcodeScanned(body)
+        }
       }
     }
```
